Thirdweb Integration - Complete Token, NFT & Smart Contract Management

soloking1412 opened the pull request:
## Description
This PR adds comprehensive Thirdweb integration to rsk-cli, enabling users to deploy, manage, and interact with ERC20 tokens, ERC721 NFTs, and arbitrary smart contracts on the Rootstock blockchain.

### Features Added

#### ERC20 Token Management
- **Deploy ERC20 tokens** with custom name and symbol
- **Mint tokens** to specified addresses
- **Transfer tokens** between addresses
- **Check balances** for any address

#### ERC721 NFT Management
- **Deploy NFT collections** with custom metadata
- **Mint NFTs** with name, description, and image URL
- **Transfer NFTs** between addresses
- **Check NFT balances** with detailed metadata display

#### Arbitrary Contract Deployment
- **Deploy any smart contract** using ABI and bytecode files
- **Support for constructor arguments**
- **Interactive prompts** for missing parameters

#### IPFS Storage Integration
- **Upload files** to IPFS using Thirdweb storage
- **Download files** from IPFS with automatic file type detection
- **Multiple gateway support** for reliable downloads

### Security & UX Features
- **Secure storage** of API keys and private keys in encrypted `rootstock-wallet.json`
- **Interactive prompts** for all commands with validation
- **Consistent emoji usage** for better visual feedback
- **Progress indicators** and transaction confirmations
- **Comprehensive error handling** with helpful messages

### Usage Examples

```bash
# Deploy ERC20 token
rsk-cli thirdweb erc20 --testnet --name "My Token" --symbol "MTK"

# Mint tokens
rsk-cli thirdweb mint --testnet --address "CONTRACT_ADDRESS" --to "RECIPIENT" --amount "1000"

# Deploy NFT collection
rsk-cli thirdweb erc721 --testnet --name "My Collection" --symbol "MC"

# Deploy arbitrary contract
rsk-cli thirdweb deploy-custom --testnet --abi "abi.json" --bytecode "bytecode.bin"

# Upload to IPFS
rsk-cli thirdweb ipfs --upload "file.jpg"
```

### Technical Implementation
- **Thirdweb SDK integration** for seamless blockchain interactions
- **Modular command structure** for maintainability
- **Type-safe implementation** with proper error handling
- **Automatic balance checking** before deployments
- **Transaction confirmation** with explorer links

### Dependencies Added
- `@thirdweb-dev/sdk`: Core Thirdweb SDK
- `commander`: CLI framework
- `inquirer`: Interactive prompts
- `chalk`: Colored output
- `ora`: Progress spinners

### Testing
- ✅ Tested on Rootstock testnet and mainnet
- ✅ Verified all token operations (deploy, mint, transfer, balance)
- ✅ Confirmed NFT operations with metadata
- ✅ Tested arbitrary contract deployment
- ✅ Validated IPFS upload/download operations
- ✅ Tested interactive mode for all commands

ezequiel-rodriguez commented:
Hey @soloking1412 ,
Thank you for your contribution — it looks very thorough!

I haven’t had the chance to go through the entire PR in detail yet, as it's quite extensive, but I wanted to share some early feedback:

I noticed that in several commands, the output is missing emojis. For example, in src/utils/thirdwebHelper.ts you're using emojis. It's important to follow the existing convention across commands and include emojis where appropriate.

Currently, only ERC-20 and ERC-721 contracts are supported for deployment. It would be great if you could add support for deploying arbitrary contract types as well.

Thanks again!
Looking forward to reviewing the rest in more detail.

soloking1412 commented:
> Hey @soloking1412 , Thank you for your contribution — it looks very thorough!
> 
> I haven’t had the chance to go through the entire PR in detail yet, as it's quite extensive, but I wanted to share some early feedback:
> 
> I noticed that in several commands, the output is missing emojis. For example, in src/utils/thirdwebHelper.ts you're using emojis. It's important to follow the existing convention across commands and include emojis where appropriate.
> 
> Currently, only ERC-20 and ERC-721 contracts are supported for deployment. It would be great if you could add support for deploying arbitrary contract types as well.
> 
> Thanks again! Looking forward to reviewing the rest in more detail.

Thank you for the feedback! I've already addressed both points:
1. Emoji consistency: Fixed all Thirdweb commands to include consistent emojis throughout (✅, 🔧, 📄, etc.) following the existing patterns in the codebase.
2. Arbitrary contract deployment: Added deploy-custom command that supports deploying any smart contract using ABI and bytecode files with constructor arguments.

scguaquetam commented:
Hello, we have found details which need to be fixed:

1. The Thirdweb SDK implementation requires packages which produces 13 vulnerabilities , please upgrade/fix it to avoid these vulnerabilities.

<img width="416" height="53" alt="image" src="https://github.com/user-attachments/assets/5acb2a42-c74c-4ae0-ba92-3ed4a92cbcb6" />

2. Please, fix the conflicts and take into consideration that we added other features to the several commands + MCP integration, so be sure everything is working fine with your new features and everything is also documented in the Readme.
3. The current feature requires to pass private key directly on the console, instead of using the stored wallets (for reference do the same flow as transfer module)
4. Even the reading functions requires Private key (does not make sense)
5. The fix made on the error message on the previous check is okay, but for the other functions, there are also not formatted errors like this, please check error logs (you can see reference in other modules):
<img width="1152" height="586" alt="image" src="https://github.com/user-attachments/assets/a64c425a-a411-4760-abbf-59d3e0148baf" />

scguaquetam commented on bin/index.ts:
```diff
@@ -57,7 +53,7 @@ const program = new Command();
   .name("rsk-cli")
   .description("CLI tool for interacting with Rootstock blockchain")
-  .version("1.2.1", "-v, --version", "Display the current version");
+  .version("1.1.0", "-v, --version", "Display the current version");
```
this is wrong

scguaquetam replied:
sorry, dont pay attention to this
